src/v/cloud_roles/aws_refresh_impl.cc

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_roles/aws_refresh_impl.h"
+
+#include "s3/client.h"
+
+struct ec2_response_schema {
+    static constexpr std::string_view expiry = "Expiration";
+    static constexpr std::string_view access_key_id = "AccessKeyId";
+    static constexpr std::string_view secret_access_key = "SecretAccessKey";
+    static constexpr std::string_view session_token = "Token";
+};
+
+cloud_roles::aws_refresh_impl::aws_refresh_impl(
+  seastar::sstring api_host,
+  uint16_t api_port,
+  s3::aws_region_name region,
+  ss::abort_source& as,
+  retry_params retry_params)
+  : refresh_credentials::impl(
+    std::move(api_host), api_port, std::move(region), as, retry_params) {}
+
+ss::future<cloud_roles::api_response>
+cloud_roles::aws_refresh_impl::fetch_credentials() {
+    if (unlikely(!_role)) {
+        vlog(clrl_log.info, "initializing role name");
+        auto response = co_await fetch_role_name();
+        // error while fetching the role, make caller handle it
+        if (std::holds_alternative<api_request_error>(response)) {
+            co_return response;
+        }
+
+        iobuf_const_parser parser(std::get<iobuf>(response));
+        _role.emplace(parser.read_string(parser.bytes_left()));
+
+        vlog(clrl_log.info, "fetched iam role name [{}]", *_role);
+        if (_role->empty()) {
+            // TODO (abhijat) create a new error kind for bad system state
+            co_return api_request_error{
+              .reason = "empty role name set on instance",
+              .error_kind = api_request_error_kind::failed_abort};
+        }
+    }
+
+    if (_role->empty()) {
+        vlog(
+          clrl_log.error,
+          "IAM role name not populated, cannot fetch credentials");
+        // TODO (abhijat) create a new error kind for bad system state
+        co_return api_request_error{
+          .reason = "missing IAM role name",
+          .error_kind = api_request_error_kind::failed_retryable};
```

Comment:
I'm not sure it is something you want to change, but it would be pretty easy and then you could get rid of the std::variant and have more idiomatic code. see consensus.h for an example where most apis return ss::future<result<T>> and then in src/v/raft/errc.h all the error codes are defined. this is the preferred way to handle this situation.

Reply:
There are upsides to this variant based approach, such as:

1. In a hierarchy of calls, error codes (probably, unless I am missing something) do not carry enough information to be able to tie errors to a specific level of calls. eg in this PR, one level of calls can return a network error. when that level succeeds, a higher level of calls can return a JSON parse error. With errors as types arranged in a tree with variants, the compiler helps us so we cannot return a parse error from the network call, and I cannot return a network error from the json parsing code.
2. Errors as types allow attaching extra context to the error, for example if the json parse succeeded but one of the required fields was missing, we can highlight which field was missing.

The downsides are that `std::variant` is a little clumsy as it does not support pattern matching on results, resulting in a lot of boilerplate where the call result is handled. Additionally in my current implementation the top level variant is not so much a tree as a flat collection of error kinds. I should probably fix that.

I do see the advantages of the errc patter, it looks a lot cleaner and is easier to extend for new error kinds etc. If the errc pattern is used in the code base I will switch to that for consistency. I can migrate the error types to error codes. 

Would it be ok to punt this to the next PR for IAM roles? If not I can try doing the conversion now.

Reply:
> Would it be ok to punt this to the next PR for IAM roles? If not I can try doing the conversion now.

of course. maintaining consistency is important and unless there is a compelling reason not to, we should probably maintain the status quo. but for this first version of the PR the variant version looks safe.